Replace ironclad dependency with cl-isaac

samuel-hunter opened the pull request:
Ironclad is a meaty dependency. This system takes the longest time on my
computer to compile my web application, and I've read comments online
that it adds ~19MB RAM usage in a lisp image.

This change switches ironclad with the single-purpose library `sha1` to
minimize lack's footprint, and the library `cl-isaac` for a
cryptographic PRNG for random session generation.

samuel-hunter commented:
Quicklisp's dist `2021-10-20` updated sha1 -- this is now doable!

fukamachi commented:
On a different topic, the `cl:random` is not quite _random_ and it will cause a conflict of session IDs on a large number of requests. So, I just rewrote it with `ironclad:random-data`. Oops.

samuel-hunter commented:
Hi @fukamachi, I've found that [cl-isaac](https://github.com/thephoeron/cl-isaac) looks like a good crypto PRNG to replace ironclad's random. If you're interested in this PR, I'll go ahead and replace `ironclad:random-data`.

fukamachi commented:
Happy to hear that! Sounds good to me.

samuel-hunter committed:
Replace ironclad dependency with cl-isaac

Ironclad is a meaty dependency. This system takes the longest time on my
computer to compile my web application, and I've read comments online
that it adds ~19MB RAM usage in a lisp image.

This change switches ironclad with the single-purpose library `cl-isaac`
for a cryptographic PRNG for random session generation.

samuel-hunter commented:
The commit's amended with cl-isaac. Thanks for taking a look at this!
On the aside, I have a similar PR for the websocket-driver: https://github.com/fukamachi/websocket-driver/pull/58

samuel-hunter commented:
I took a second look at this. Since it already depends on `cl+ssl`, isaac isn't needed, and we can draw PRNG from there instead. No extra libraries may be needed.

samuel-hunter commented:
On the aside, `cl+ssl:random-bytes` looks like an equally suitable alternative.

fukamachi commented on src/util.lisp:
```diff
@@ -81,7 +78,14 @@
-  (ironclad:byte-array-to-hex-string
-    (ironclad:random-data 20)))
+  (format nil "~64,'0x" (#+:X86-64 isaac:rand-bits-64
+                         #-:X86-64 isaac:rand-bits-32
```
This raises a following error on my M1 Mac:

```
; caught ERROR:
;   READ error during COMPILE-FILE:
;
;     Symbol "RAND-BITS-32" not found in the CL-ISAAC package.
;
;       Line: 91, Column: -1, File-Position: 3523
;
;       Stream: #<SB-INT:FORM-TRACKING-STREAM for "file /home/fukamachi/Programs/lib/lack/src/util.lisp" {1005F58D23}>
COMPILE-FILE-ERROR while
compiling #<CL-SOURCE-FILE "lack-util" "src/util">
   [Condition of type UIOP/LISP-BUILD:COMPILE-FILE-ERROR]
```

The function name for 32bit machines should be [Isaac:rand-bits](https://github.com/thephoeron/cl-isaac/blob/938318303d31a7be886ce3b68b427b087d519d90/packages.lisp#L16), maybe?

```suggestion
                         #-:X86-64 isaac:rand-bits
```

fukamachi replied:
And also, it would be great to treat ARM64 as 64bit.
This needs to fix cl-isaac, though.

```
CL-USER> *features*
(:SWANK :QUICKLISP :QUICKLISP-SUPPORT-HTTPS :ASDF3.3 :ASDF3.2 :ASDF3.1 :ASDF3
 :ASDF2 :ASDF :OS-UNIX :NON-BASE-CHARS-EXIST-P :ASDF-UNICODE :ROS.INIT :ARM64
 :GENCGC :64-BIT :ANSI-CL :COMMON-LISP :ELF :IEEE-FLOATING-POINT :LINUX
 :LITTLE-ENDIAN :PACKAGE-LOCAL-NICKNAMES :SB-CORE-COMPRESSION :SB-LDB
 :SB-PACKAGE-LOCKS :SB-THREAD :SB-UNICODE :SBCL :UNIX)
```

samuel-hunter replied:
Good catch, fixed!

> And also, it would be great to treat ARM64 as 64bit.

cl-isaac only supports 64-bit when the `x86-64` feature is flagged: https://github.com/thephoeron/cl-isaac/blob/260cb349d4251d838d6cb6c7bda66177f7a75480/cl-isaac.asd#L29

fukamachi reviewed (changes requested):
Thanks for your patch!
Almost great, but I found a minor architecture problem. Could you confirm it?

samuel-hunter committed:
Quickfix generate-random-id on non x86-64 systems

fukamachi commented on src/util.lisp:
```diff
@@ -81,7 +78,14 @@
   "Generates a random token."
-  (ironclad:byte-array-to-hex-string
-    (ironclad:random-data 20)))
+  (format nil "~64,'0x" (#+:X86-64 isaac:rand-bits-64
+                         #-:X86-64 isaac:rand-bits
+                         *isaac-ctx* 512)))
```
I found another problem.
This generates a 128-char string, though the previous version generates 40-chars.

It could break some compatibility.
For example, the following is a validation code in lack.middleware.session that assumes the session ID is a lower-cased 40-char string.

https://github.com/fukamachi/lack/blob/156ef010d5a90e629107b44661221b26b507d7ef/src/middleware/session/state.lisp#L18-L22

fukamachi replied:
If it's okay with you, this snippet works for me.

```suggestion
  (format nil "~(~40,'0x~)" (#+:X86-64 isaac:rand-bits-64
                             #-:X86-64 isaac:rand-bits
                             *isaac-ctx* 160)))
```

samuel-hunter replied:
LGTM, applied

samuel-hunter committed:
Shorten `generate-random-id` result to 40 chars

The current function generates a 128-char string, though the previous generates 40-chars.

It could break some compatibility.
For example, the following is a validation code in lack.middleware.session that assumes the session ID is a lower-cased 40-char string.

This commit reduces the generated bits to 160 to shorten the session Id to 40 chars

Co-authored-by: Eitaro Fukamachi <e.arrows@gmail.com>

fukamachi reviewed (approved):
LGTM!

fukamachi commented:
Thanks a lot!
